core/templates/services/utils.service.ts

```diff
@@ -128,6 +128,46 @@ export class UtilsService {
         element.offsetHeight < element.scrollHeight);
     }
   }
+
+  // Determines whether the URL is pointing to a page on the Oppia site.
+  getSafeReturnUrl(urlString: string): string {
```

Comment:
`getSafeReturnUrl('//evil.com')` returns `//evil.com`, what may allow redirect to external websites:
1. `new URL('//evil.com')` (line 144) throws exception but validation continues
2. `new URL(urlString, document.baseURI);` (line 157) succeeds 
3. `urlString.charAt(0) !== '/'`(line 163) condition is satisfied

Only relative URLs should be considered safe.
If, redirecting to external websites is a requirement, then known host names should be kept in an AllowList and the `urlString` checked against that list.

Reply:
Gotcha, thanks! Redirecting to external websites is not a requirement, so we're fine with only trusting relative paths. Do you know of an API that can detect them securely?

Reply:
I am not an Angular expert but would some of the [available Router methods](https://angular.io/api/router/Router) do the job e.g. [`navigateByUrl()`](https://angular.io/api/router/Router#navigatebyurl)?

I see that [Angular Router has a `parseUrl()` method](https://angular.io/api/router/Router#parseurl): it may worth a look at.

If I had to do it in play JavaScript/TypeScript, probably I would go with something simple (please don't use it as is without further testing):

```typescript
getSafeReturnUrl(urlString: string): string {
  const anchor = document.createElement('a');
  anchor.href = urlString;

  return (anchor.host === window.location.host)? urlString : '/';
}
```
| Input | Output |
| :------ | :-------- |
| `//evil.com` | `/` |
| `https://evil.com` | `/` |
| `javascript:alert(1)` | `/` |
| `data:,Hello%2C%20World%21` | `/` |
| `/login` | `/login` |
| `/` | `/` |

Reply:
This is super helpful, thanks Paulo!

I know @srijanreddy98 has been promoting the usage of Angular router. Srijan, do you know of a way to protect against these "fake return URLs" using angular/routers?

Reply:
For now, I've added more test cases and a simple additional check since the approach you provided doesn't protect us from URL-encoded attacks. For now, we're at least a little more secure (I can't find an exhaustive list of checks so I'll consider this resolved once the test cases seem "good enough").